File: Behavior/ChainOfResponsibility/TypeScript/chainOfResponsibility.ts
abstract class Handler {
    protected nextHandler: Handler | null = null;

    public setNext(handler: Handler): Handler {
        this.nextHandler = handler;
        return handler;
    }

    public handle(request: string): string {
        if (this.nextHandler) {
            return this.nextHandler.handle(request);
        }

        return '';
    }
}

class ConcreteHandler1 extends Handler {
    public handle(request: string): string {
        if (request === 'request1') {
            return 'ConcreteHandler1 handled the request';
        } else if (this.nextHandler) {
            return this.nextHandler.handle(request);
        }

        return '';
    }
}

class ConcreteHandler2 extends Handler {
    public handle(request: string): string {
        if (request === 'request2') {
            return 'ConcreteHandler2 handled the request';
        } else if (this.nextHandler) {
            return this.nextHandler.handle(request);
        }

        return '';
    }
}

// 使用示例
const handler1 = new ConcreteHandler1();
const handler2 = new ConcreteHandler2();

handler1.setNext(handler2);

console.log(handler1.handle('request1'));  // Output: ConcreteHandler1 handled the request
console.log(handler1.handle('request2'));  // Output: ConcreteHandler2 handled the request

File: Behavior/ChainOfResponsibility/Cpp/chainOfResponsibility.cpp
#include <iostream>
#include <vector>

// 抽象处理者类
class Handler {
public:
    virtual ~Handler() {}
    void SetSuccessor(Handler* successor) {
        successor_ = successor;
    }
    virtual void HandleRequest(int request) = 0;

protected:
    Handler* successor_;
};

// 具体处理者类1
class ConcreteHandler1 : public Handler {
public:
    virtual void HandleRequest(int request) override {
        if (request < 10) {
            std::cout << "ConcreteHandler1 handled request " << request << std::endl;
        } else if (successor_ != nullptr) {
            successor_->HandleRequest(request);
        }
    }
};

// 具体处理者类2
class ConcreteHandler2 : public Handler {
public:
    virtual void HandleRequest(int request) override {
        if (request >= 10 && request < 20) {
            std::cout << "ConcreteHandler2 handled request " << request << std::endl;
        } else if (successor_ != nullptr) {
            successor_->HandleRequest(request);
        }
    }
};

// 客户端代码
int main() {
    ConcreteHandler1 handler1;
    ConcreteHandler2 handler2;

    handler1.SetSuccessor(&handler2);

    std::vector<int> requests = {2, 5, 14, 22, 18, 3, 27, 20};

    for (int request : requests) {
        handler1.HandleRequest(request);
    }

    return 0;
}
